src/async/parallel.ts

```diff
@@ -9,30 +9,46 @@ type WorkItemResult<K> = {
  * Executes many async functions in parallel. Returns the results from
  * all functions as an array. After all functions have resolved, if
  * any errors were thrown, they are rethrown in an instance of
- * AggregateError.
+ * AggregateError. The operation can be aborted by passing optional AbortSignal,
+ * which will throw an Error if aborted.
  *
  * @see https://radashi.js.org/reference/async/parallel
  * @example
  * ```ts
  * // Process images concurrently, resizing each image to a standard size.
+ * const abortController = new AbortController();
  * const images = await parallel(2, imageFiles, async (file) => {
  *   return await resizeImage(file)
- * })
+ * }, abortController)
+ *
+ * // To abort the operation:
+ * // abortController.abort()
  * ```
  * @version 12.1.0
  */
 export async function parallel<T, K>(
   limit: number,
   array: readonly T[],
   func: (item: T) => Promise<K>,
+  signal?: AbortSignal,
 ): Promise<K[]> {
   const work = array.map((item, index) => ({
     index,
     item,
   }))
   // Process array items
-  const processor = async (res: (value: WorkItemResult<K>[]) => void) => {
+  const processor = async (
+    res: (value: WorkItemResult<K>[]) => void,
+    rej: (error: any) => void,
+  ) => {
     const results: WorkItemResult<K>[] = []
+    signal?.addEventListener(
+      'abort',
+      () => rej(new Error('Operation aborted')),
+      {
+        once: true,
+      },
+    )
```

Comment:
I don't think this event listener is necessary. If you're using an AbortSignal with `parallel`, you'll be using it within your `func` callback, which will naturally propagate the `AbortError` when aborted.

Reply:
In my opinion it is necessary, because there are a couple of edge cases where code from previous comment will fail:

```diff
-    while (true) {
+    while (!signal?.aborted) {
```

- If `AbortController.abort` is called before the invocation of `parallel` and `signal?.aborted` is true it will get stuck in an infinite loop

We can fix this by doing something like this:

```ts
 while (true) {
      const next = work.pop()
      if (!next || options?.signal?.aborted) {
        return res(results)
      }
      const [error, result] = await tryit(func)(next.item)
      results.push({
        error,
        result: result as K,
        index: next.index,
      })
    }
```

But, if `AbortController.abort` is called between iterations, the `AbortError` will be skipped in the `results` array and the `parallel` will not throw `AggregationError`

And in case `func` executes for a considerable amount of time I think it will be more favorable to just throw `parallel` immediately than wait till it finishes its execution